x/pairing/keeper/migrations.go

```diff
@@ -24,33 +26,78 @@ func (m Migrator) MigrateVersion2To3(ctx sdk.Context) error {
 	return nil
 }
 
-// MigrateVersion3To4 fix delegation total in the stake entries
-func (m Migrator) MigrateVersion3To4(ctx sdk.Context) error {
+func (m Migrator) MigrateVersion4To5(ctx sdk.Context) error {
 	entries := m.keeper.epochStorageKeeper.GetAllStakeEntriesCurrent(ctx)
-	epoch := m.keeper.epochStorageKeeper.GetCurrentNextEpoch(ctx)
-	for _, e := range entries {
-		delegations, err := m.keeper.dualstakingKeeper.GetProviderDelegators(ctx, e.Address, epoch)
+	providerMap := map[string][]*epochstoragetypes.StakeEntry{}
+
+	// map the providers
+	for i := range entries {
+		providerMap[entries[i].Address] = append(providerMap[entries[i].Address], &entries[i])
+	}
+
+	// iterate over each provider
+	for address, entries := range providerMap {
+		metadata := epochstoragetypes.ProviderMetadata{Provider: address}
+
+		// find the bigest vault
+		var biggestVault *epochstoragetypes.StakeEntry
+		for i := range entries {
+			if metadata.Description.Moniker == "" {
+				metadata.Description = entries[i].Description
+			}
+			e := entries[i]
+			if e.Vault != e.Address {
+				if biggestVault == nil {
+					biggestVault = e
+				} else if biggestVault.Stake.Amount.LT(e.Stake.Amount) {
+					biggestVault = e
+				}
+			}
+		}
+
+		// if no vault was found the vault is the address
+		if biggestVault != nil {
+			metadata.Vault = biggestVault.Vault
+		} else {
+			metadata.Vault = address
+		}
+
+		// get all delegations and sum
+		delegations, err := m.keeper.dualstakingKeeper.GetProviderDelegators(ctx, metadata.Provider)
 		if err != nil {
-			utils.LavaFormatError("failed getting provider delegators at MigrateVersion3To4", err, utils.LogAttr("provider", e.Address))
-			continue
+			panic("ahahahaha")
 		}
 
-		delegateTotal := sdk.ZeroInt()
+		metadata.TotalDelegations = sdk.NewCoin(m.keeper.stakingKeeper.BondDenom(ctx), sdk.ZeroInt())
 		for _, d := range delegations {
-			if e.Address == d.Delegator || e.Vault == d.Delegator || d.ChainID != e.Chain {
-				continue
+			if d.Delegator != metadata.Vault {
+				metadata.TotalDelegations = metadata.TotalDelegations.Add(d.Amount)
 			}
-			delegateTotal = delegateTotal.Add(d.Amount.Amount)
 		}
-		if !e.DelegateTotal.Amount.Equal(delegateTotal) {
-			fmt.Println("fixing delegate total for", e.Address, e.Chain)
 
-			e.DelegateTotal.Amount = delegateTotal
-			if e.TotalStake().LT(m.keeper.specKeeper.GetMinStake(ctx, e.Chain).Amount) {
-				e.Freeze()
+		// fix entries with different vaults
+		// count self delegations
+		TotalSelfDelegation := sdk.ZeroInt()
+		for _, e := range entries {
+			if e.Vault != metadata.Vault {
+				fmt.Println(address)
+				biggestVault.Stake = biggestVault.Stake.SubAmount(sdk.NewInt(1))
+				e.Stake.Amount = sdk.OneInt()
```

Comment:
i don't get this, you take from the biggest vault 1ulava and put here this will create a mismatch between the vault stake and the vault delegation on the biggestVault chain and add to the stake entry of other non vault chains

Reply:
resolve